### ISSUE TITLE
The "defines" js.Build option fails silently for almost anything besides single alphanumeric words

### ISSUE THREAD

**Issue by LandGod** [gohugoio/hugo#8040]
### What version of Hugo are you using (`hugo version`)?
<pre>
$ hugo version
Hugo Static Site Generator v0.79.0/extended windows/amd64 BuildDate: unknown
</pre>

### Does this issue reproduce with the latest release?
Yes.

### Summary
When using the `"defines"` option with `js.Build` many special characters such as `-`, `<`, and `!`, as well as any internal spaces, will cause `js.Build` to return a `nil` reference without throwing any error. The build will later fail when you try to insert the link to the resource into your layout file with `$myBuiltJS.RelPermalink`.

EG:
```
Error: Error building site: failed to render pages: render of "home" failed: "C:\Users\Dan\Documents\Programming\HugoTest\quickstart\layouts\index.html:47:31": execute of template failed: template: index.html:47:31: executing "main" at <$testJs.RelPermalink>: error calling RelPermalink: runtime error: invalid memory address or nil pointer dereference
```

The issue is two-fold here:
1. `js.Build` just fails so easily that it's really hard to use it for anything other than the very most basic substitutions. I assume (and hope) that this is not the intended behavior. 
2. Because no error is thrown from `js.Build` itself, the resulting build failure is very confusing and there is no easy way to tell that a) The problem even came from "defines" in the first place, and b) what about the provided value to "defines" caused it to fail. 

I spun up a fresh Hugo repo to demonstrate this issue. Check out the `index.html` layout to see various exampled of passing and failing values for "defines". (You'll have to un-comment each one individually to see.)

https://github.com/LandGod/HugoTestBed

**Comment by bep** [comment]
This smells like an upstream issue (in the ESBuild project) -- we just pass these defines on as-is -- and I would somehow expect them to validate these.

**Comment by evanw** [comment]
I was curious to see what this problem was so I took a look. It looks like the code for processing errors from esbuild tries to access `msg.Location.File` without checking for `nil` first:

https://github.com/gohugoio/hugo/blob/212e5e554284bc9368e52a512ed09be5a0224d3e/resources/resource_transformers/js/build.go#L100-L101

I assume this then leads to a crash in Hugo's code: `runtime error: invalid memory address or nil pointer dereference`. The `Location` field in esbuild's API is deliberately a pointer to a struct instead of a plain struct because it can be `nil`. This corresponds to an error that isn't associated with a file. In this case the error is with the API options so there is no file location and the location is `nil`.

> When using the `"defines"` option with `js.Build` many special characters such as `-`, `<`, and `!`, as well as any internal spaces, will cause `js.Build` to return a `nil` reference without throwing any error.

The define feature is documented here: https://esbuild.github.io/api/#define. The key must be a valid JavaScript identifier or identifier chain, and the value must either be a valid JavaScript identifier or a JSON value.

**Comment by jmooring** [comment]
This was resolved in v0.81.0 with https://github.com/gohugoio/hugo/commit/a1fe552fc9e622a15010a94281f604eb85bebd84.
